File: without-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java
package com.example.app;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.lang.reflect.Type;
import java.util.Map;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import com.stripe.android.ApiResultCallback;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.PaymentIntentResult;
import com.stripe.android.Stripe;
import com.stripe.android.exception.APIConnectionException;
import com.stripe.android.exception.APIException;
import com.stripe.android.exception.AuthenticationException;
import com.stripe.android.exception.InvalidRequestException;
import com.stripe.android.model.ConfirmPaymentIntentParams;
import com.stripe.android.model.PaymentIntent;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.PaymentMethodCreateParams;
import com.stripe.android.model.StripeIntent;
import com.stripe.android.view.CardInputWidget;

import org.jetbrains.annotations.NotNull;


public class CheckoutActivity extends AppCompatActivity {
    /**
     * To run this app, you'll need to first run the sample server locally.
     * Follow the "How to run locally" instructions in the root directory's README.md to get started.
     * Once you've started the server, open http://localhost:4242 in your browser to check that the
     * server is running locally.
     * After verifying the sample server is running locally, build and run the app using the
     * Android emulator.
     */
    // 10.0.2.2 is the Android emulator's alias to localhost
    private String backendUrl = "http://10.0.2.2:4242/";
    private OkHttpClient httpClient = new OkHttpClient();
    private Stripe stripe;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_checkout);
        loadPage();
    }

    private void loadPage() {
        // Clear the card widget
        CardInputWidget cardInputWidget = findViewById(R.id.cardInputWidget);
        cardInputWidget.clear();

        // Load Stripe publishable key from the server
        Request request = new Request.Builder()
                .url(backendUrl + "stripe-key")
                .get()
                .build();
        httpClient.newCall(request)
                .enqueue(new Callback() {
                    @Override
                    public void onFailure(Call call, IOException e) {
                        runOnUiThread(() -> {
                            Context applicationContext = getApplicationContext();
                            Toast.makeText(applicationContext, "Error: " + e.toString(), Toast.LENGTH_LONG).show();
                        });
                    }

                    @Override
                    public void onResponse(Call call, final Response response) throws IOException {
                        if (!response.isSuccessful()) {
                            runOnUiThread(() -> {
                                Context applicationContext = getApplicationContext();
                                Toast.makeText(applicationContext, "Error: " + response.toString(), Toast.LENGTH_LONG).show();
                            });
                        } else {
                            Gson gson = new Gson();
                            Type type = new TypeToken<Map<String, String>>(){}.getType();
                            Map<String, String> map = gson.fromJson(response.body().string(), type);
                            String stripePublishableKey = map.get("publishableKey");

                            runOnUiThread(() -> {
                                // Use the publishable key from the server to initialize the Stripe instance.
                                Context applicationContext = getApplicationContext();
                                PaymentConfiguration.init(applicationContext, stripePublishableKey);
                                stripe = new Stripe(applicationContext, stripePublishableKey);

                                // Hook up the pay button to the card widget and stripe instance
                                Button payButton = findViewById(R.id.payButton);
                                payButton.setOnClickListener((View view) -> {
                                    pay();
                                });
                            });
                        }
                    }
                });
    }

    private void pay() {
        CardInputWidget cardInputWidget = findViewById(R.id.cardInputWidget);
        PaymentMethodCreateParams params = cardInputWidget.getPaymentMethodCreateParams();

        if (params == null) {
            return;
        }
        stripe.createPaymentMethod(params, new ApiResultCallback<PaymentMethod>() {
            @Override
            public void onSuccess(@NonNull PaymentMethod result) {
                // Create and confirm the PaymentIntent by calling the sample server's /pay endpoint.
                pay(result.id, null);
            }

            @Override
            public void onError(@NonNull Exception e) {

            }
        });
    }

    private void pay(String paymentMethod, String paymentIntent) {
        WeakReference<Activity> weakActivity = new WeakReference<>(this);
        MediaType mediaType = MediaType.get("application/json; charset=utf-8");
        String json;
        if (paymentMethod != null) {
            json = "{"
                    + "\"paymentMethodId\":" + "\"" + paymentMethod + "\","
                    + "\"currency\":\"usd\","
                    + "\"items\":["
                    + "{\"id\":\"photo_subscription\"}"
                    + "]"
                    + "}";
        } else {
            json = "{"
                    + "\"paymentIntentId\":" +  "\"" + paymentIntent + "\""
                    + "}";
        }
        RequestBody body = RequestBody.create(json, mediaType);
        Request request = new Request.Builder()
                .url(backendUrl + "pay")
                .post(body)
                .build();
        httpClient.newCall(request)
                .enqueue(new Callback() {
                    @Override
                    public void onFailure(Call call, IOException e) {
                        runOnUiThread(() -> {
                            Context applicationContext = getApplicationContext();
                            Toast.makeText(applicationContext, "Error: " + e.toString(), Toast.LENGTH_LONG).show();
                        });
                    }

                    @Override
                    public void onResponse(Call call, final Response response) throws IOException {
                        if (!response.isSuccessful()) {
                            runOnUiThread(() -> {
                                Context applicationContext = getApplicationContext();
                                Toast.makeText(applicationContext, "Error: " + response.toString(), Toast.LENGTH_LONG).show();
                            });
                        } else {
                            Gson gson = new Gson();
                            Type type = new TypeToken<Map<String, String>>(){}.getType();
                            Map<String, String> responseMap = gson.fromJson(response.body().string(), type);

                            String error = responseMap.get("error");
                            String paymentIntentClientSecret = responseMap.get("clientSecret");
                            String requiresAction = responseMap.get("requiresAction");

                            if (error != null) {
                                displayAlert(weakActivity.get(), "Error", error.toString(), false);
                            } else if (paymentIntentClientSecret != null && requiresAction != "true") {
                                displayAlert(weakActivity.get(), "Payment succeeded", paymentIntentClientSecret, true);
                            } else if (paymentIntentClientSecret != null && requiresAction == "true") {
                                runOnUiThread(() -> {
                                    stripe.authenticatePayment(weakActivity.get(), paymentIntentClientSecret);
                                });
                            }
                        }
                    }
                });
    }

    private void displayAlert(Activity activity, String title, String message, boolean restartDemo) {
        runOnUiThread(() -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(activity);
            builder.setTitle(title);
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            builder.setMessage(message);
            if (restartDemo) {
                builder.setPositiveButton("Restart demo", (DialogInterface dialog, int index) -> {
                    loadPage();
                });
            } else {
                builder.setPositiveButton("Ok", null);
            }
            AlertDialog dialog = builder.create();
            dialog.show();
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        WeakReference<Activity> weakActivity = new WeakReference<>(this);

        // Handle the result of stripe.confirmPayment
        stripe.onPaymentResult(requestCode, data, new ApiResultCallback<PaymentIntentResult>() {
            @Override
            public void onSuccess(@NonNull PaymentIntentResult result) {
                PaymentIntent paymentIntent = result.getIntent();
                PaymentIntent.Status status = paymentIntent.getStatus();
                if (status == PaymentIntent.Status.Succeeded) {
                    // Payment completed successfully
                    runOnUiThread(() -> {
                        if (weakActivity.get() != null) {
                            Activity activity = weakActivity.get();
                            Gson gson = new GsonBuilder().setPrettyPrinting().create();
                            displayAlert(activity, "Payment completed", gson.toJson(paymentIntent), true);
                        }
                    });
                } else if (status == PaymentIntent.Status.RequiresPaymentMethod) {
                    // Payment failed – allow retrying using a different payment method
                    runOnUiThread(() -> {
                        if (weakActivity.get() != null) {
                            Activity activity = weakActivity.get();
                            displayAlert(activity, "Payment failed", paymentIntent.getLastPaymentError().message, false);
                        }
                    });
                } else if (status == PaymentIntent.Status.RequiresConfirmation) {
                    // After handling a required action on the client, the status of the PaymentIntent is
                    // requires_confirmation. You must send the PaymentIntent ID to your backend
                    // and confirm it to finalize the payment. This step enables your integration to
                    // synchronously fulfill the order on your backend and return the fulfillment result
                    // to your client.
                    pay(null, paymentIntent.getId());
                }
            }

            @Override
            public void onError(@NonNull Exception e) {
                // Payment request failed – allow retrying using the same payment method
                runOnUiThread(() -> {
                    if (weakActivity.get() != null) {
                        Activity activity = weakActivity.get();
                        displayAlert(activity, "Error", e.toString(), false);
                    }
                });
            }
        });
    }
}

File: without-webhooks/client/android-kotlin/app/src/main/java/com/example/app/CheckoutActivity.kt
package com.example.app

import java.io.IOException
import java.lang.ref.WeakReference

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody

import com.google.gson.GsonBuilder
import org.json.JSONObject

import com.stripe.android.ApiResultCallback
import com.stripe.android.PaymentIntentResult
import com.stripe.android.Stripe
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.StripeIntent
import com.stripe.android.view.CardInputWidget


class CheckoutActivity : AppCompatActivity() {

    /**
     * To run this app, you'll need to first run the sample server locally.
     * Follow the "How to run locally" instructions in the root directory's README.md to get started.
     * Once you've started the server, open http://localhost:4242 in your browser to check that the
     * server is running locally.
     * After verifying the sample server is running locally, build and run the app using the
     * Android emulator.
     */
    // 10.0.2.2 is the Android emulator's alias to localhost
    private val backendUrl = "http://10.0.2.2:4242/"
    private val httpClient = OkHttpClient()
    private lateinit var publishableKey: String
    private lateinit var stripe: Stripe

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_checkout)
        startCheckout()
    }

    private fun displayAlert(activity: Activity?, title: String, message: String, restartDemo: Boolean = false) {
        if (activity == null) {
            return
        }
        runOnUiThread {
            val builder = AlertDialog.Builder(activity!!)
            builder.setTitle(title)
            builder.setMessage(message)
            if (restartDemo) {
                builder.setPositiveButton("Restart demo") { _, _ ->
                    val cardInputWidget =
                        findViewById<CardInputWidget>(R.id.cardInputWidget)
                    cardInputWidget.clear()
                    startCheckout()
                }
            }
            else {
                builder.setPositiveButton("Ok", null)
            }
            val dialog = builder.create()
            dialog.show()
        }
    }

    private fun startCheckout() {
        val weakActivity = WeakReference<Activity>(this)
        // Load Stripe key from the server
        val request = Request.Builder()
            .url(backendUrl + "stripe-key")
            .get()
            .build()
        httpClient.newCall(request)
            .enqueue(object: Callback {
                override fun onFailure(call: Call, e: IOException) {
                    displayAlert(weakActivity.get(), "Failed to load page", "Error: $e")
                }

                override fun onResponse(call: Call, response: Response) {
                    if (!response.isSuccessful) {
                        displayAlert(weakActivity.get(), "Failed to load page", "Error: $response")
                    } else {
                        val responseData = response.body?.string()
                        var json = JSONObject(responseData)

                        // The response from the server includes the Stripe public key and
                        // PaymentIntent details.
                        publishableKey = json.getString("publishableKey")

                        // Use the public key from the server to initialize the Stripe instance.
                        stripe = Stripe(applicationContext, publishableKey)
                    }
                }
            })

        val payButton: Button = findViewById(R.id.payButton)
        payButton.setOnClickListener {
            pay()
        }
    }

    private fun pay() {
        val weakActivity = WeakReference<Activity>(this)
        // Collect card details on the client
        val cardInputWidget =
            findViewById<CardInputWidget>(R.id.cardInputWidget)
        val params = cardInputWidget.paymentMethodCreateParams
        if (params == null) {
            return
        }
        stripe.createPaymentMethod(params, object : ApiResultCallback<PaymentMethod> {
            // Create PaymentMethod failed
            override fun onError(e: Exception) {
                displayAlert(weakActivity.get(), "Payment failed", "Error: $e")
            }
            override fun onSuccess(result: PaymentMethod) {
                // Create a PaymentIntent on the server with a PaymentMethod
                print("Created PaymentMethod")
                pay(result.id, null)
            }
        })
    }

    // Create or confirm a PaymentIntent on the server
    private fun pay(paymentMethod: String?, paymentIntent: String?) {
        val weakActivity = WeakReference<Activity>(this)
        var json = ""
        if (!paymentMethod.isNullOrEmpty()) {
            json = """
                {
                    "useStripeSdk":true,
                    "paymentMethodId":"$paymentMethod",
                    "currency":"usd",
                    "items": [
                        {"id":"photo_subscription"}
                    ]
                }
                """
        }
        else if (!paymentIntent.isNullOrEmpty()) {
            json = """
                {
                    "paymentIntentId":"$paymentIntent"
                }
                """
        }
        // Create a PaymentIntent on the server
        val mediaType = "application/json; charset=utf-8".toMediaType()
        val body = json.toRequestBody(mediaType)
        val request = Request.Builder()
            .url(backendUrl + "pay")
            .post(body)
            .build()
        httpClient.newCall(request)
            .enqueue(object: Callback {
                override fun onFailure(call: Call, e: IOException) {
                    displayAlert(weakActivity.get(), "Payment failed", "Error: $e")
                }

                override fun onResponse(call: Call, response: Response) {
                    // Request failed
                    if (!response.isSuccessful) {
                        displayAlert(weakActivity.get(), "Payment failed", "Error: $response")
                    } else {
                        val responseData = response.body?.string()
                        var responseJson = JSONObject(responseData)
                        val payError: String? = responseJson.optString("error")
                        val clientSecret: String? = responseJson.optString("clientSecret")
                        val requiresAction: Boolean? = responseJson.optBoolean("requiresAction")
                        // Payment failed
                        if (payError != null && payError.isNotEmpty()) {
                            displayAlert(weakActivity.get(), "Payment failed", "Error: $payError")
                        }
                        // Payment succeeded
                        else if ((clientSecret != null && clientSecret.isNotEmpty())
                            && (requiresAction == null || requiresAction == false)) {
                            displayAlert(weakActivity.get(), "Payment succeeded", "$clientSecret", restartDemo = true)
                        }
                        // Payment requires additional actions
                        else if ((clientSecret != null && clientSecret.isNotEmpty())
                            && requiresAction == true) {
                            runOnUiThread {
                                if (weakActivity.get() != null) {
                                    val activity = weakActivity.get()!!
                                    stripe.authenticatePayment(activity, clientSecret)
                                }
                            }
                        }
                    }
                }
            })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val weakActivity = WeakReference<Activity>(this)

        // Handle the result of stripe.authenticatePayment
        stripe.onPaymentResult(requestCode, data, object : ApiResultCallback<PaymentIntentResult> {
            override fun onSuccess(result: PaymentIntentResult) {
                val paymentIntent = result.intent
                val status = paymentIntent.status
                if (status == StripeIntent.Status.Succeeded) {
                    val gson = GsonBuilder().setPrettyPrinting().create()
                    displayAlert(weakActivity.get(), "Payment succeeded", gson.toJson(paymentIntent), restartDemo = true)
                } else if (status == StripeIntent.Status.RequiresPaymentMethod) {
                    // Payment failed – allow retrying using a different payment method
                    displayAlert(weakActivity.get(), "Payment failed", paymentIntent.lastPaymentError!!.message ?: "")
                }
                // After handling a required action on the client, the status of the PaymentIntent is
                // requires_confirmation. You must send the PaymentIntent ID to your backend
                // and confirm it to finalize the payment. This step enables your integration to
                // synchronously fulfill the order on your backend and return the fulfillment result
                // to your client.
                else if (status == StripeIntent.Status.RequiresConfirmation) {
                    print("Re-confirming PaymentIntent after handling a required action")
                    pay(null, paymentIntent.id)
                }
                else {
                    displayAlert(weakActivity.get(), "Payment status unknown", "unhandled status: $status", restartDemo = true)
                }
            }

            override fun onError(e: Exception) {
                // Payment request failed – allow retrying using the same payment method
                displayAlert(weakActivity.get(), "Payment failed", e.toString())
            }
        })
    }

}
